### PR TITLE
filestream: validate input id on startup

### PR DESCRIPTION
**Pull request by AndersonQ**
During startup filebeat now validates the filestream inputs and fails to start if there are inputs without ID or with duplicated IDs. Duplicated IDs might cause data duplication therefore now it's mandatory to have unique and non-empty ID for each filestream input. 

## Proposed commit message
```
filestream: validate input id on startup

During startup filebeat now validates the filestream inputs and fails to start if there are with duplicated IDs. Duplicated IDs might cause data duplication therefore now it's mandatory to have unique for each filestream input. 

Disruptive User Impact
Impact: Users who have not configured unique IDs for their filestream inputs will need to update their configurations to include unique IDs for each input. Failure to do so will prevent Filebeat from starting.

Previously: Filebeat would only log an error if filestream would find inputs with duplicated IDs, potentially leading to data duplication.

Now: Filebeat will fail to start if any filestream input has a duplicated ID.
```

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact

**Impact:** Users who have not configured unique IDs for their filestream inputs will need to update their configurations to include unique IDs for each input. Failure to do so will prevent Filebeat from starting.

**Previously:** Filebeat would only log an error if filestream would find inputs had missing or duplicated IDs, potentially leading to data duplication.

**Now:** Filebeat will fail to start if any filestream input lacks an ID or has a duplicated ID.

## How to test this PR locally

 - build filebeat
```shell
cd filebeat
go build .
```
 - use the following as `filebeat.yml`. Change as necessary to simulate" only unique IDs, valid empty ID, duplicated IDs, duplicated empty ID.
```yaml
filebeat.inputs:
  - type: filestream
    enabled: true
    paths:
      - /tmp/noID-1/*.log
  - type: filestream
    enabled: true
    paths:
      - /tmp/noID-2/*.log
  - type: filestream
    id: duplicated-id
    enabled: true
    paths:
      - /tmp/duplicated-id-1/*.log
  - type: filestream
    id: duplicated-id
    enabled: true
    paths:
      - /tmp/duplicated-id-2/*.log

output.discard.enabled: true
logging:
  level: info
  metrics:
    enabled: false
```

- run filebeat:
```shell
./filebeat -e 2>&1 | grep message
```

### verify filebeat:
  - will exit with error
  - will produce a log with all the invalid input configuration:
```json
{"log.level":"error","@timestamp":"2024-11-25T10:44:19.062+0100","log.logger":"filestream","log.origin":{"function":"github.com/elastic/beats/v7/filebeat/input/filestream.ValidateInputIDs","file.name":"filestream/config.go","file.line":174},"message":"filestream inputs with duplicated IDs","service.name":"filebeat","inputs":[{"enabled":true,"paths":["/tmp/noID-1/*.log"],"type":"filestream"},{"enabled":true,"paths":["/tmp/noID-2/*.log"],"type":"filestream"},{"enabled":true,"id":"duplicated-id","paths":["/tmp/duplicated-id-1/*.log"],"type":"filestream"},{"enabled":true,"id":"duplicated-id","paths":["/tmp/duplicated-id-2/*.log"],"type":"filestream"}],"ecs.version":"1.6.0"}
```
  - will produce a log pointing out the issues:
```json
{"log.level":"error","@timestamp":"2024-11-25T10:44:19.062+0100","log.origin":{"function":"github.com/elastic/beats/v7/filebeat/beater.(*Filebeat).Run","file.name":"beater/filebeat.go","file.line":297},"message":"invalid filestream configuration: filestream inputs validation error: filestream inputs with duplicated IDs: \"\",\"duplicated-id\"","service.name":"filebeat","ecs.version":"1.6.0"}
Exiting: filestream inputs validation error: filestream inputs with duplicated IDs: "","duplicated-id"
```

## Related issues

- Relates #40540

## Logs

```json
{"log.level":"info","@timestamp":"2024-11-25T10:44:19.062+0100","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/cmd/instance.(*Beat).launch","file.name":"instance/beat.go","file.line":770},"message":"filebeat start running.","service.name":"filebeat","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-11-25T10:44:19.062+0100","log.logger":"filestream","log.origin":{"function":"github.com/elastic/beats/v7/filebeat/input/filestream.ValidateInputIDs","file.name":"filestream/config.go","file.line":174},"message":"filestream inputs with duplicated IDs","service.name":"filebeat","inputs":[{"enabled":true,"paths":["/tmp/noID-1/*.log"],"type":"filestream"},{"enabled":true,"paths":["/tmp/noID-2/*.log"],"type":"filestream"},{"enabled":true,"id":"duplicated-id","paths":["/tmp/duplicated-id-1/*.log"],"type":"filestream"},{"enabled":true,"id":"duplicated-id","paths":["/tmp/duplicated-id-2/*.log"],"type":"filestream"}],"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-11-25T10:44:19.062+0100","log.origin":{"function":"github.com/elastic/beats/v7/filebeat/beater.(*Filebeat).Run","file.name":"beater/filebeat.go","file.line":297},"message":"invalid filestream configuration: filestream inputs validation error: filestream inputs with duplicated IDs: \"\",\"duplicated-id\"","service.name":"filebeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-11-25T10:44:19.062+0100","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/cmd/instance.(*Beat).launch","file.name":"instance/beat.go","file.line":779},"message":"filebeat stopped.","service.name":"filebeat","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-11-25T10:44:19.062+0100","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/cmd/instance.handleError","file.name":"instance/beat.go","file.line":1594},"message":"Exiting: filestream inputs validation error: filestream inputs with duplicated IDs: \"\",\"duplicated-id\"","service.name":"filebeat","ecs.version":"1.6.0"}
Exiting: filestream inputs validation error: filestream inputs with duplicated IDs: "","duplicated-id"
```

### CODE REVIEW

File: filebeat/input/filestream/config.go [hunk 1]
```diff
@@ -142,3 +144,78 @@ func (c *config) Validate() error {
...
+
+func collectOffendingInputs(duplicates []string, empty []*conf.C, ids map[string][]*conf.C) []map[string]interface{} {
+	var cfgs []map[string]interface{}
+
+	if len(empty) > 0 {
+		for _, cfg := range empty {
+			toJson := map[string]interface{}{}
+			err := cfg.Unpack(&toJson)
+			if err != nil {
+				toJson["emptyID"] = fmt.Sprintf("failes to umpack config: %v", err)
```
Comment (mauri870) [inline filebeat/input/filestream/config.go:203]:
```suggestion
				toJson["emptyID"] = fmt.Sprintf("failed to unpack config: %v", err)
```

File: filebeat/input/filestream/config.go [hunk 2]
```diff
@@ -142,3 +144,78 @@ func (c *config) Validate() error {
...
+			cfgs = append(cfgs, toJson)
+		}
+	}
+
+	for _, id := range duplicates {
+		for _, dupcfgs := range ids[id] {
+			toJson := map[string]interface{}{}
+			err := dupcfgs.Unpack(&toJson)
+			if err != nil {
+				toJson[id] = fmt.Sprintf("failes to umpack config: %v", err)
```
Comment (mauri870) [inline filebeat/input/filestream/config.go:214]: Same here

**Review by mauri870 — APPROVED** [review]
Codewise looks good, left some comments about some minor spelling issues.

I tested locally following the issue description and confirmed that it validates the ids properly. 

LGTM.

File: filebeat/input/filestream/config.go [hunk 3]
```diff
@@ -142,3 +144,78 @@ func (c *config) Validate() error {
...
+		if err != nil {
+			return fmt.Errorf("failed to unpack filestream input configuration: %w", err)
+		}
+		if fsInput.Type == "filestream" {
+			if fsInput.ID == "" {
+				empty = append(empty, input)
+				continue
+			}
+			ids[fsInput.ID] = append(ids[fsInput.ID], input)
+			if len(ids[fsInput.ID]) > 1 {
```
Comment (rdner) [inline filebeat/input/filestream/config.go:171]: I don't see any value in collecting the same ID multiple times. If we have the same ID used in 6 different inputs it would end up in the error message 5 times.

```suggestion
			if len(ids[fsInput.ID]) == 2 {
```

File: filebeat/input/filestream/config.go [hunk 4]
```diff
@@ -142,3 +144,78 @@ func (c *config) Validate() error {
...
+	for _, id := range duplicates {
+		for _, dupcfgs := range ids[id] {
+			toJson := map[string]interface{}{}
+			err := dupcfgs.Unpack(&toJson)
+			if err != nil {
+				toJson[id] = fmt.Sprintf("failed to unpack config: %v", err)
+			}
+			cfgs = append(cfgs, toJson)
+		}
+	}
```
Comment (rdner) [inline filebeat/input/filestream/config.go:218]: See my previous comment about `duplicates`.
Without the change above we would collect the same list of configs multiple times here, I don't think this is what's desired.

File: filebeat/input/filestream/config_test.go [hunk 5]
```diff
@@ -30,3 +36,158 @@ func TestConfigValidate(t *testing.T) {
...
+				require.Len(t, logs, 1, "there should be only one log entry")
+
+				got, err := json.Marshal(logs[0].ContextMap()["inputs"])
+				require.NoError(t, err, "could not marshal duplicated IDs inputs")
+				assert.Equal(t, want, string(got))
+
+			},
+		},
+		{
+			name: "duplicated IDs",
```
Comment (rdner) [inline filebeat/input/filestream/config_test.go:82]: We need to add a test case when you have more than 2 duplicates (5?) and the error message does not repeat the same input more than once.

**Comment by cmacknz** [comment]
> if there are inputs without ID 

An empty ID is fine, the problem is if there are two inputs with an empty ID. Agent allows the empty ID as a valid ID. This avoids breaking the configuration unnecessarily where there is only a single filestream input with no ID.

**Review by rdner — APPROVED** [review]
All my comments have been addressed.
We might want to treat an empty ID as a regular ID value as @cmacknz pointed out. So, we can apply the duplication check to that as well.

**Comment by AndersonQ** [comment]
I removed the bit forbidding an empty ID. It was already checking for multiple empty IDs and including them in the list of invalid inputs.

**Comment by rdner** [comment]
@AndersonQ looks like the tests need adjusting?

```
=== FAIL: filebeat/tests/integration TestFilestreamValidationPreventsFilebeatStart/empty_ID (10.12s)
    filestream_test.go:180:
        	Error Trace:	/opt/buildkite-agent/builds/bk-agent-prod-gcp-1732528923926044677/elastic/filebeat/libbeat/tests/integration/framework.go:540
        	            				/opt/buildkite-agent/builds/bk-agent-prod-gcp-1732528923926044677/elastic/filebeat/filebeat/tests/integration/filestream_test.go:180
        	Error:      	Condition never satisfied
        	Test:       	TestFilestreamValidationPreventsFilebeatStart/empty_ID
        	Messages:   	Filebeat did log a validation error
```

File: filebeat/tests/integration/filestream_test.go [hunk 6]
```diff
@@ -105,3 +108,122 @@ func TestFilestreamCleanInactive(t *testing.T) {
...
+
+			// Write configuration file and start Filebeat
+			filebeat.WriteConfigFile(tc.cfg)
+			filebeat.Start()
+
+			// Wait for error log
+			filebeat.WaitForLogs(
+				"filestream inputs validation error",
+				10*time.Second,
+				"Filebeat did log a validation error")
```
Comment (belimawr) [inline filebeat/tests/integration/filestream_test.go:183]: should it be "Filebeat did not log a validation error"?
```suggestion
				"Filebeat did not log a validation error")
```

File: filebeat/tests/integration/filestream_test.go [hunk 7]
```diff
@@ -105,3 +108,122 @@ func TestFilestreamCleanInactive(t *testing.T) {
...
+
+	// Write configuration file and start Filebeat
+	filebeat.WriteConfigFile(cfg)
+	filebeat.Start()
+
+	// Wait for error log
+	filebeat.WaitForLogs(
+		"Input 'filestream' starting",
+		10*time.Second,
+		"Filebeat did log a validation error")
```
Comment (belimawr) [inline filebeat/tests/integration/filestream_test.go:228]: Maybe something else, like "filebeat did not start the Filestream input"
```suggestion
		"Filebeat did not start the Filestream input")
```

**Comment by belimawr** [comment]
> > if there are inputs without ID
> 
> An empty ID is fine, the problem is if there are two inputs with an empty ID. Agent allows the empty ID as a valid ID. This avoids breaking the configuration unnecessarily where there is only a single filestream input with no ID.

Honestly, as we're already doing a breaking change to enforce ID validation, I rather not allow empty IDs, it is a corner case and Elastic-Agent/Fleet already adds IDs to all inputs, so I don't see a real need for it. It only makes for an extra corner case to maintain.

**Comment by belimawr** [comment]
@AndersonQ, @cmacknz, Should we also cover the case of dynamic inputs like:
- Input reload is enabled and inputs are read from `inputs.d` + `filebeat.yml`
- Filebeat autodiscover
- Filebeat running under Elastic-Agent

I like the way it's implemented here that gives this quick and precise feedback to the use, however it does not cover all cases.

I believe we should, at least, also cover the `inputs.d` if Filebeat is setup to read inputs from it, even if it's just at startup.

**Comment by AndersonQ** [comment]
> Honestly, as we're already doing a breaking change to enforce ID validation, I rather not allow empty IDs

Agreed, but also what was approved in the breaking change committee was indeed to enforce the uniqueness of IDs. Also considering the agent already allows a single empty ID, I agree with Craig it's fine to have an empty ID as long as it's just one. 

> @AndersonQ, @cmacknz, Should we also cover the case of dynamic inputs like:
> 
> * Input reload is enabled and inputs are read from `inputs.d` + `filebeat.yml`
> * Filebeat autodiscover
> * Filebeat running under Elastic-Agent
> 
> I like the way it's implemented here that gives this quick and precise feedback to the use, however it does not cover all cases.

Those are different things. Here we're tacking the start up and early feedback.

In order to completely prevent an input with duplicated ID, we'd need to change it to return an error instead of just logging it. Which happens in a completely different part of the code. Having it on another issue and PR if we really want to do that is a better option.

File: filebeat/input/filestream/config.go [hunk 8]
```diff
@@ -142,3 +144,57 @@ func (c *config) Validate() error {
...
+			ID   string `config:"id"`
+			Type string `config:"type"`
+		}{}
+		err := input.Unpack(&fsInput)
+		if err != nil {
+			return fmt.Errorf("failed to unpack filestream input configuration: %w", err)
+		}
+		if fsInput.Type == "filestream" {
+			ids[fsInput.ID] = append(ids[fsInput.ID], input)
+			if len(ids[fsInput.ID]) == 2 {
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply filebeat/input/filestream/config.go:166]: yes, but it just needs to go to the list once. If it uses `> 1`, it'll add the id twice in the list and later when collecting the offending configs, it'll add the same config as many times as the ID is in the list. The `duplicates` list is just a slice, it does not guarantee uniqueness.

**Comment by cmacknz** [comment]
> Filebeat running under Elastic-Agent

Since 8.6.0 Elastic Agent has required that all inputs have a unique ID, and it considers an empty ID a unique ID. The constraint is that two inputs can't have the same ID, not that the ID has to be non-empty.

https://github.com/elastic/elastic-agent/blob/434986ed9fa6fbed6257557a4c6773d519742f66/pkg/component/component.go#L734-L745

```go
func hasDuplicate(outputsMap map[string]outputI, id string) bool {
	for _, o := range outputsMap {
		for _, i := range o.inputs {
			for _, j := range i {
				if j.id == id {
					return true
				}
			}
		}
	}
	return false
}
```

If we were designing this from scratch we'd probably forbid empty IDs, but allowing empty IDs solves the underlying functional problem without unnecessarily people who started from our default configurations that didn't specify IDs a while ago. We could log a warning that IDs should be explicit if you want.

> Input reload is enabled and inputs are read from inputs.d + filebeat.yml
> Filebeat autodiscover

+1 to handling these as well, autodiscovery with the ID composed from variable substitutions with non-unique values is a common way people were ending up with duplicated logs on K8s. The original issue doesn't put a constraint on the execution environment :)

This doesn't need to be fixed in this PR though, it can be done as a separate PR, but I don't think we can close https://github.com/elastic/beats/issues/40540 until it's done, otherwise we have implemented "sometimes require unique IDs" and not "always require unique IDs".

File: filebeat/input/filestream/config.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (belimawr) [reply filebeat/input/filestream/config.go:166]: I see it now, thanks for clarifying.

File: filebeat/input/filestream/config.go [hunk 9]
```diff
@@ -142,3 +144,57 @@ func (c *config) Validate() error {
 
 	return nil
 }
+
+// ValidateInputIDs checks all filestream inputs to ensure all have an ID and
+// the ID is unique. If there is any empty or duplicated ID, it logs an error
+// containing the offending input configurations and returns an error containing
+// the duplicated IDs.
+func ValidateInputIDs(inputs []*conf.C, logger *logp.Logger) error {
+	ids := make(map[string][]*conf.C)
```
Comment (belimawr) [inline filebeat/input/filestream/config.go:153]: [Suggestion]
Rename this map to something that better describe its contents. It is a map of inputs (or configs) grouped by ID, so `ids` does not seem to be the better description to it.

**Comment by mauri870** [comment]
Reviewed again, LGTM.

**Comment by AndersonQ** [comment]
@cmacknz, regarding

> This doesn't need to be fixed in this PR though, it can be done as a separate PR, but I don't think we can close https://github.com/elastic/beats/issues/40540 until it's done, otherwise we have implemented "sometimes require unique IDs" and not "always require unique IDs".

Whereas I agree this PR only enforces unique IDs at startup, it's exactly what https://github.com/elastic/beats/issues/40540. Perhaps it'd be better to create another issue, or change the whole issue to be indeed about fielstream requiring unique IDs and add 2 sub issues, one with the current content of https://github.com/elastic/beats/issues/40540 and another for enforce uniqueness at runtime.

Also I believe when requiring uniqueness at runtime we should not crash filebeat, I believe it's better for the users to log an error and not start the offending input. Thinking about a standalone filebeat, which might have different modules running at the same time, I don't think it'd be good for the users to have filebeat crashing because a possible data duplication. Better to have just the offending issue not running.

File: filebeat/input/filestream/config.go [hunk 10]
```diff
@@ -142,3 +144,57 @@ func (c *config) Validate() error {
 
 	return nil
 }
+
+// ValidateInputIDs checks all filestream inputs to ensure all have an ID and
+// the ID is unique. If there is any empty or duplicated ID, it logs an error
+// containing the offending input configurations and returns an error containing
+// the duplicated IDs.
```
Comment (rdner) [inline filebeat/input/filestream/config.go:151]: This description is not true anymore because we allow empty IDs.

```suggestion
// ValidateInputIDs checks all filestream inputs to ensure all input IDs are unique.
// If there is a duplicated ID, it logs an error containing the offending input
// configurations and returns an error containing the duplicated IDs.
// An empty ID is treated as a valid ID value. Multiple empty IDs are not unique.
```

File: filebeat/input/filestream/config.go (hunk 8 above)
Comment (rdner) [inline filebeat/input/filestream/config.go:166]:
```suggestion
			// we want to add each config to duplicates only once, hence only when it hits `2`
			if len(ids[fsInput.ID]) == 2 {
```

**Review by rdner — APPROVED** [review]
2 suggestions for code readability, the new changes look good to me.

**Comment by cmacknz** [comment]
> Also I believe when requiring uniqueness at runtime we should not crash filebeat, I believe it's better for the users to log an error and not start the offending input. Thinking about a standalone filebeat, which might have different modules running at the same time, I don't think it'd be good for the users to have filebeat crashing because a possible data duplication. Better to have just the offending issue not running.

My guide for how this should be have is to test what Elastic Agent does and then follow its behavior because this problem is completely solved there.

When dynamic variable substitution is involved, we don't need to fail at runtime because templated inputs with the same ID end up only creating a single input in agent. I will open an issue for you to either confirm Filebeat also does this or implement that behavior based on what you find Filebeat does in that situation.
